File: Mar-20/08-03-20/java/01-palindrome-count-1.java
import java.util.*;
import java.io.*;


class TestClass {
    public static void main(String args[] ) throws Exception {
        BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
        //Scanner input = new Scanner(System.in);
        String str = input.readLine();
        int n = str.length();
        int pali[][] = new int[n][n];
        for (int i = 0; i < n; i++)
        {
            pali[i][i] = 1;
            if (i-1 >=0)
                pali[i][i-1] = 1;
        }
        int cnt = 0;
        
        for (int k = 2; k <= n; k++)
        {
            for (int i = 0; i < n - k +1; i++)
            {
                int j = i + k -1;
                int diff = j - i + 1;
                if (str.charAt(i) == str.charAt(j))
                {
                    if (diff % 2 != 0)
                    {
                        if (pali[i+1][j-1] > 0 )
                        {
                            pali[i][j] = 1;
                            cnt += 1;
                        }
                    }
                    else
                    {
                        if (pali[i+1][j-1] > 0 && (pali[i][j-1] > 0 || pali[i+1][j] > 0))
                        {
                            pali[i][j] = 1;
                            cnt += 1;
                        }
                        
                    }
                }
            }
        }
        System.out.println(n+cnt);
        
    }
}

File: Mar-20/08-03-20/python/01-palindrome-count-1.py
data = input()
n = len(data)
pali = [[0 for i in range(n)] for j in range(n)] 
 
 
for i in range(0, n):
    pali[i][i] = 1
    if (i-1 >= 0):
        pali[i][i-1] = 1
 
cnt = 0
 
for k in range(2, n+1):
    for i in range(0, n-k+1):
        j = i + k - 1
        diff = j - i + 1
        
        if (data[i] == data[j]):
            if (diff % 2 != 0):
                if (pali[i+1][j-1]):
                    pali[i][j] = 1
                    cnt = cnt + 1
            else:
                if (pali[i+1][j-1] and (pali[i][j-1] or pali[i+1][j])):
                    pali[i][j] = 1
                    cnt = cnt + 1
 
# for i in range(0, n):
#     for j in range(0, n):
#         print(pali[i][j], end = " ")
#     print()
 
print(n + cnt)
